sklearn/tree/_criterion.pyx

```diff
@@ -1330,3 +1330,122 @@ cdef class FriedmanMSE(MSE):
                                self.weighted_n_node_samples))
+
+
+cdef class Poisson(RegressionCriterion):
+    """Half Poisson deviance as impurity criterion.
+
+    Poisson deviance = 2/n * sum(y_true * log(y_true/y_pred) + y_pred - y_true)
+
+    Note that the deviance is >= 0, and since we have `y_pred = mean(y_true)`
+    at the leaves, one always has `sum(y_pred - y_true) = 0`. It remains the
+    implemented impurity:
+        1/n * sum(y_true * log(y_true/y_pred)
+    """
+    # FIXME in 0.25:
+    # min_impurity_split with default = 0 forces us to use a non-negative
+    # impurity like the Poisson deviance. Without this restriction, one could
+    # throw away the 'constant' term sum(y_true * log(y_true)) and just use
+    # Poisson loss = - 1/n * sum(y_true * log(y_pred))
+    #              = - 1/n * sum(y_true * log(mean(y_true))
+    #              = - mean(y_true) * log(mean(y_true))
+    # With this trick (used in proxy_impurity_improvement()), as for MSE,
+    # children_impurity would only need to go over left xor right split, not
+    # both. This could be faster.
+
+    cdef double node_impurity(self) nogil:
+        """Evaluate the impurity of the current node.
+
+        Evaluate the Poisson criterion as impurity of the current node,
+        i.e. the impurity of samples[start:end]. The smaller the impurity the
+        better.
+        """
+        return self.poisson_loss(self.start, self.end, self.sum_total,
+                                 self.weighted_n_node_samples)
+
+    cdef double proxy_impurity_improvement(self) nogil:
+        """Compute a proxy of the impurity reduction.
+
+        This method is used to speed up the search for the best split.
+        It is a proxy quantity such that the split that maximizes this value
+        also maximizes the impurity improvement. It neglects all constant terms
+        of the impurity decrease for a given split.
+
+        The absolute impurity improvement is only computed by the
+        impurity_improvement method once the best split has been found.
+
+        Poisson proxy is:
+            - 1/n * sum(y_i * log(y_pred)) = -mean(y_i) * log(mean(y_i))
+        """
+        cdef SIZE_t k
+        cdef double proxy_impurity_left = 0.0
+        cdef double proxy_impurity_right = 0.0
+        cdef double y_mean_left = 0.
+        cdef double y_mean_right = 0.
+
+        for k in range(self.n_outputs):
+            if (self.sum_left[k] <= 0) or (self.sum_right[k] <= 0):
+                # Poisson loss does not allow non-positive predictions. We
+                # therefore forbid splits that have child nodes with
+                # sum(y_i) <= 0.
+                return -INFINITY
+            else:
+                y_mean_left = self.sum_left[k] / self.weighted_n_left
+                y_mean_right = self.sum_right[k] / self.weighted_n_right
+                proxy_impurity_left -= y_mean_left * log(y_mean_left)
```

Comment:
Is it on purpose to not use `xlogy` here and use it below?

Reply:
Here, it can't happen that y_mean_... is zero. `xlogx = x * log(y)` is particularly useful in situations where `x=0` can happen with the expected overall result being zero.

Reply:
OK I see, cool.